src/devices/sound/dmadac.cpp

```diff
@@ -108,9 +112,33 @@ void dmadac_sound_device::transfer(int channel, offs_t channel_spacing, offs_t f
 		if (j != total_frames)
 			logerror("dmadac_transfer: buffer overrun (short %d frames)\n", total_frames - j);
 	}
+}
 
-	// FIXME: this line has rotted and can no longer compile - it should be fixed and uncommented or removed
-	//LOG("dmadac_transfer - %d samples, %d effective, %d in buffer\n", total_frames, int(total_frames * double(DEFAULT_SAMPLE_RATE) / dmadac[first_channel].frequency), dmadac[first_channel].curinpos - dmadac[first_channel].curoutpos);
+void dmadac_sound_device::transfer(int channel, offs_t channel_spacing, offs_t frame_spacing, offs_t total_frames, int32_t *data)
+{
+	int j;
+
+	/* loop over all channels and accumulate the data */
+	constexpr stream_buffer::sample_t sample_scale = 1.0 / double(std::numeric_limits<int32_t>::max());
+	if (m_enabled)
+	{
+		int maxin = (m_bufout + BUFFER_SIZE - 1) % BUFFER_SIZE;
+		int32_t *src = data + channel * channel_spacing;
+		int curin = m_bufin;
+
+		/* copy the data */
+		for (j = 0; j < total_frames && curin != maxin; j++)
+		{
+			m_buffer[curin] = stream_buffer::sample_t(*src) * sample_scale;
+			curin = (curin + 1) % BUFFER_SIZE;
+			src += frame_spacing;
+		}
+		m_bufin = curin;
+
+		/* log overruns */
+		if (j != total_frames)
+			logerror("dmadac_transfer: buffer overrun (short %d frames)\n", total_frames - j);
+	}
```

Comment:
There seems to be growing copy-pasta here.  Is it possible to template it somehow to reduce duplication without losing performance?

Reply:
Yeah, I wasn't a fan of the duplication going on and considered templates as well. Ultimately I went with what I considered the "safe" option in my initial code because I don't have the strongest grasp of things like performance implications from templates.

I gave it a shot in this commit: https://github.com/mamedev/mame/pull/8434/commits/751d26b5a751a170649d9a10f158443173a73502

I did search for any code I could find that used the `transfer` function directly, and I only found usages of the `int16_t` buffer version, never the `stream_buffer::sample_t` buffer version. If you pass in a `stream_buffer::sample_t` buffer then it'll still do the scaling part in this templated code which I am a little iffy about due to floating point. But it doesn't seem to be used so it may not be an actual issue.

Performance-wise, there weren't any noticeable performance changes in the games I gave a test.